modules/platforms/cpp/core/include/ignite/impl/cluster/cluster_node_impl.h

```diff
@@ -44,27 +45,189 @@ namespace ignite
                 /**
                  * Constructor used to create new instance.
                  *
-                 * @param reader Binary reader.
+                 * @param mem Memory to read Cluster Node.
                  */
-                ClusterNodeImpl(binary::BinaryReaderImpl& reader);
+                ClusterNodeImpl(common::concurrent::SharedPointer<interop::InteropMemory> mem);
 
                 /**
                  * Destructor.
                  */
                 ~ClusterNodeImpl();
 
+                /**
+                 * Get collection of addresses this node is known by.
+                 *
+                 * @return Collection of addresses this node is known by.
+                 */
+                std::vector<std::string> GetAddresses();
+
+                /**
+                 * Get a node attribute.
+                 *
+                 * @param name Node attribute name.
+                 * @return Node attribute.
+                 *
+                 * @throw IgniteError in case of attribute name does not exist
+                 * or if template type is not compatible with attribute.
+                 */
+                template<typename T>
+                T GetAttribute(std::string name)
+                {
+                    if (attrs.Get()->find(name) == attrs.Get()->end())
+                    {
+                        const char* msg = "The is no Cluster Node attribute with name requested";
+                        throw IgniteError(IgniteError::IGNITE_ERR_ILLEGAL_ARGUMENT, msg);
+                    }
+
+                    interop::InteropInputStream stream(mem.Get());
+                    stream.Position(attrs.Get()->find(name)->second);
+
+                    binary::BinaryReaderImpl reader(&stream);
+
+                    return reader.ReadObject<T>();
+                }
+
+                /**
+                 * Get collection of all Cluster Node attributes names.
+                 *
+                 * @return Node attributes names collection.
+                 */
+                std::vector<std::string> GetAttributes();
+
+                /**
+                 * Get Cluster Node consistent ID.
+                 *
+                 * @return Cluster Node consistent ID.
+                 */
+                template<typename T>
+                T GetConsistentId()
```

Comment:
We should return `std::string` here, or it will not usable for user.

Reply:
It is not possible because sometimes it could be Guid.
It is possible to convert Guid to string, but I am not sure that it is suitable for user.

Reply:
Yes it is. AFAIK, this is the only way it is used in Java right now - everyone just call `.toString()` on returned method - because how else you can really use it?

The most widespread usage of this method is comparison and strings are perfectly comparable just as well as GUIDs.

Reply:
Now the std::string returned in all cases.